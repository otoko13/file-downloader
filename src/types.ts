export type FileStatus = "available" | "scheduled";

export type File = {
  name: string;
  device: string;
  path: string;
  status: FileStatus;
};
