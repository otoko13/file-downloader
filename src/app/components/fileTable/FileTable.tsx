import { File } from "@/types";
import classNames from "classnames";
import { FunctionComponent, useCallback, useMemo, useState } from "react";
import { CheckedState } from "../checkbox/Checkbox";
import styles from "./fileTable.module.scss";
import FileTableHeader from "./FileTableHeader";
import FileTableRow from "./FileTableRow";

interface FileTableProps {
  className?: string;
  files: File[];
}

const FileTable: FunctionComponent<FileTableProps> = ({ className, files }) => {
  const [selectedFiles, setSelectedFiles] = useState<File[]>([]);
  const availableFiles = useMemo(
    () => files.filter((file) => file.status === "available"),
    [files]
  );

  const handleSelectAllClicked = () => {
    const filesToSelect =
      selectedFiles.length === availableFiles.length ? [] : availableFiles;
    setSelectedFiles(filesToSelect);
  };

  const getSelectAllCheckedState = (): CheckedState => {
    if (selectedFiles.length === availableFiles.length) {
      return "checked";
    }
    if (selectedFiles.length === 0) {
      return "unchecked";
    }
    return "partial";
  };

  const handleFileSelectClicked = (file: File) => {
    const indexInSelected = selectedFiles.findIndex(
      (selectedFile) => selectedFile.path === file.path
    );
    if (indexInSelected < 0) {
      setSelectedFiles((files) => [...files, file]);
      return;
    }
    const updatedSelection = selectedFiles.filter(
      (selectedFile) => selectedFile.path !== file.path
    );
    setSelectedFiles(updatedSelection);
  };

  const handleDownloadClicked = useCallback(() => {
    if (!selectedFiles.length) {
      return;
    }
    const alertSegments = selectedFiles
      .filter((file) => file.status === "available")
      .map((file) => `${file.path} on device ${file.device}`);
    const alertText = `Downloading: ${alertSegments.join("\n")}`;
    alert(alertText);
  }, [selectedFiles]);

  return (
    <table className={classNames(styles["file-table"], className)}>
      <FileTableHeader
        checkedState={getSelectAllCheckedState()}
        onDownloadClicked={handleDownloadClicked}
        onSelectAllClicked={handleSelectAllClicked}
        selectedFileCount={selectedFiles.length}
      />
      <tbody>
        {files.map((file) => (
          <FileTableRow
            file={file}
            key={file.path}
            selected={selectedFiles.some(
              (selectedFile) => selectedFile.path === file.path
            )}
            onClick={() => handleFileSelectClicked(file)}
          />
        ))}
      </tbody>
    </table>
  );
};

export default FileTable;
