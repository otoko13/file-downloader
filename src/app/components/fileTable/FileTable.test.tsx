import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import testFiles from "../../../testData.json";
import FileTable from "./FileTable";
import { File } from "@/types";

const files = testFiles as File[];

const clickSelectAllCheckbox = async () => {
  await userEvent.click(screen.getByLabelText("Select all"));
};

const clickRowCheckbox = async (row: number) => {
  await userEvent.click(screen.getAllByRole("checkbox")[row]);
};

const getRowCheckbox = (row: number) => screen.getAllByRole("checkbox")[row];

describe("apperance", () => {
  it("should match the snapshot", () => {
    const { container } = render(<FileTable files={files} />);
    expect(container).toMatchSnapshot();
  });
});

describe("clicking download", () => {
  let windowAlertMock: jest.Mock;

  beforeEach(() => {
    windowAlertMock = jest.fn();
    window.alert = windowAlertMock;
  });

  it("should not trigger an alert if no files selected", async () => {
    render(<FileTable files={files} />);
    await userEvent.click(
      screen.getByRole("button", { name: "Download Selected" })
    );
    expect(windowAlertMock).not.toHaveBeenCalled();
  });

  it("should trigger an alert if files are selected", async () => {
    render(<FileTable files={files} />);
    await clickRowCheckbox(2);
    await clickRowCheckbox(3);
    await userEvent.click(
      screen.getByRole("button", { name: "Download Selected" })
    );
    expect(windowAlertMock).toHaveBeenCalledWith(
      "Downloading: \\Device\\HarddiskVolume2\\Windows\\System32\\netsh.exe on device Luigi\n\\Device\\HarddiskVolume1\\Windows\\System32\\uxtheme.dll on device Peach"
    );
  });
});

describe("selecting rows", () => {
  it("should update the selected count indicator", async () => {
    render(<FileTable files={files} />);
    await clickRowCheckbox(2);
    expect(screen.getByText("Selected 1")).toBeInTheDocument();
    await clickRowCheckbox(3);
    expect(screen.getByText("Selected 2")).toBeInTheDocument();
  });

  it("should not update the selected count indicator if an unavailable row is clicked", async () => {
    render(<FileTable files={files} />);
    await clickRowCheckbox(2);
    expect(screen.getByText("Selected 1")).toBeInTheDocument();
    await clickRowCheckbox(1);
    expect(screen.getByText("Selected 1")).toBeInTheDocument();
  });

  it("should change the status of the select all checkbox", async () => {
    render(<FileTable files={files} />);
    const selectAllCheckbox = screen.getByLabelText("Select all");
    expect(selectAllCheckbox.getAttribute("aria-checked")).toBe("false");
    await clickRowCheckbox(2);
    expect(selectAllCheckbox.getAttribute("aria-checked")).toBe("mixed");
    await clickRowCheckbox(3);
    expect(selectAllCheckbox.getAttribute("aria-checked")).toBe("true");
  });

  it("should not change the status of the select all checkbox if an unavailble row is clicked", async () => {
    render(<FileTable files={files} />);
    const selectAllCheckbox = screen.getByLabelText("Select all");
    expect(selectAllCheckbox.getAttribute("aria-checked")).toBe("false");
    await clickRowCheckbox(2);
    expect(selectAllCheckbox.getAttribute("aria-checked")).toBe("mixed");
    await clickRowCheckbox(1);
    await clickRowCheckbox(4);
    await clickRowCheckbox(5);
    await clickRowCheckbox(6);
    expect(selectAllCheckbox.getAttribute("aria-checked")).toBe("mixed");
  });

  it("should allow deselection", async () => {
    render(<FileTable files={files} />);
    await clickRowCheckbox(2);
    expect(screen.getByText("Selected 1")).toBeInTheDocument();
    await clickRowCheckbox(2);
    expect(screen.getByText("Selected 0")).toBeInTheDocument();
  });
});

describe("select all checkbox", () => {
  it("should start unchecked", () => {
    render(<FileTable files={files} />);
    const selectAllCheckbox = screen.getByLabelText("Select all");
    expect(selectAllCheckbox.getAttribute("aria-checked")).toBe("false");
  });

  it("should select all available files if none are checked", async () => {
    render(<FileTable files={files} />);
    const selectAllCheckbox = screen.getByLabelText("Select all");
    await clickSelectAllCheckbox();
    expect(selectAllCheckbox.getAttribute("aria-checked")).toBe("true");
    expect(getRowCheckbox(1)).not.toBeChecked();
    expect(getRowCheckbox(2)).toBeChecked();
    expect(getRowCheckbox(3)).toBeChecked();
    expect(getRowCheckbox(4)).not.toBeChecked();
    expect(getRowCheckbox(5)).not.toBeChecked();
    expect(getRowCheckbox(6)).not.toBeChecked();
  });

  it("should select all available files if some are checked", async () => {
    render(<FileTable files={files} />);
    const selectAllCheckbox = screen.getByLabelText("Select all");
    await clickRowCheckbox(3);
    await clickSelectAllCheckbox();
    expect(selectAllCheckbox.getAttribute("aria-checked")).toBe("true");
    expect(getRowCheckbox(1)).not.toBeChecked();
    expect(getRowCheckbox(2)).toBeChecked();
    expect(getRowCheckbox(3)).toBeChecked();
    expect(getRowCheckbox(4)).not.toBeChecked();
    expect(getRowCheckbox(5)).not.toBeChecked();
    expect(getRowCheckbox(6)).not.toBeChecked();
  });

  it("should deselect all if all available files are checked", async () => {
    render(<FileTable files={files} />);
    const selectAllCheckbox = screen.getByLabelText("Select all");
    await clickSelectAllCheckbox();
    expect(selectAllCheckbox.getAttribute("aria-checked")).toBe("true");
    await clickSelectAllCheckbox();
    expect(selectAllCheckbox.getAttribute("aria-checked")).toBe("false");
    expect(getRowCheckbox(1)).not.toBeChecked();
    expect(getRowCheckbox(2)).not.toBeChecked();
    expect(getRowCheckbox(3)).not.toBeChecked();
    expect(getRowCheckbox(4)).not.toBeChecked();
    expect(getRowCheckbox(5)).not.toBeChecked();
    expect(getRowCheckbox(6)).not.toBeChecked();
  });
});
