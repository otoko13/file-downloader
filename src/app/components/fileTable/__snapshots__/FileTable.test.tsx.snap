// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apperance should match the snapshot 1`] = `
<div>
  <table
    class="file-table"
  >
    <thead
      class="file-table-header"
    >
      <tr>
        <th
          colspan="5"
        >
          <div
            class="top-header"
          >
            <button
              aria-checked="false"
              aria-label="Select all"
              class="checkbox"
              role="checkbox"
            />
            <div
              class="selected-count"
            >
              Selected 
              0
            </div>
            <button
              class="download-button"
            >
              <i
                class="bi-download download-icon"
              />
              Download Selected
            </button>
          </div>
        </th>
      </tr>
      <tr>
        <th />
        <th>
          Name
        </th>
        <th>
          Device
        </th>
        <th>
          Path
        </th>
        <th
          class="status-header"
        >
          Status
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class="file-table-row unselectable"
      >
        <td
          class="selected"
        >
          <button
            aria-checked="false"
            aria-label="Select"
            class="checkbox disabled"
            disabled=""
            role="checkbox"
          />
        </td>
        <td>
          smss.exe
        </td>
        <td>
          Mario
        </td>
        <td>
          \\Device\\HarddiskVolume2\\Windows\\System32\\smss.exe
        </td>
        <td>
          <div
            class="status-cell"
          >
            <div
              aria-hidden="true"
              aria-label="available"
              class="available-marker"
              role="img"
            />
            Scheduled
          </div>
        </td>
      </tr>
      <tr
        class="file-table-row"
      >
        <td
          class="selected"
        >
          <button
            aria-checked="false"
            aria-label="Select"
            class="checkbox"
            role="checkbox"
          />
        </td>
        <td>
          netsh.exe
        </td>
        <td>
          Luigi
        </td>
        <td>
          \\Device\\HarddiskVolume2\\Windows\\System32\\netsh.exe
        </td>
        <td>
          <div
            class="status-cell"
          >
            <div
              aria-hidden="false"
              aria-label="available"
              class="available-marker visible"
              role="img"
            />
            Available
          </div>
        </td>
      </tr>
      <tr
        class="file-table-row"
      >
        <td
          class="selected"
        >
          <button
            aria-checked="false"
            aria-label="Select"
            class="checkbox"
            role="checkbox"
          />
        </td>
        <td>
          uxtheme.dll
        </td>
        <td>
          Peach
        </td>
        <td>
          \\Device\\HarddiskVolume1\\Windows\\System32\\uxtheme.dll
        </td>
        <td>
          <div
            class="status-cell"
          >
            <div
              aria-hidden="false"
              aria-label="available"
              class="available-marker visible"
              role="img"
            />
            Available
          </div>
        </td>
      </tr>
      <tr
        class="file-table-row unselectable"
      >
        <td
          class="selected"
        >
          <button
            aria-checked="false"
            aria-label="Select"
            class="checkbox disabled"
            disabled=""
            role="checkbox"
          />
        </td>
        <td>
          aries.sys
        </td>
        <td>
          Daisy
        </td>
        <td>
          \\Device\\HarddiskVolume1\\Windows\\System32\\aries.sys
        </td>
        <td>
          <div
            class="status-cell"
          >
            <div
              aria-hidden="true"
              aria-label="available"
              class="available-marker"
              role="img"
            />
            Scheduled
          </div>
        </td>
      </tr>
      <tr
        class="file-table-row unselectable"
      >
        <td
          class="selected"
        >
          <button
            aria-checked="false"
            aria-label="Select"
            class="checkbox disabled"
            disabled=""
            role="checkbox"
          />
        </td>
        <td>
          cryptbase.dll
        </td>
        <td>
          Yoshi
        </td>
        <td>
          \\Device\\HarddiskVolume1\\Windows\\System32\\cryptbase.dll
        </td>
        <td>
          <div
            class="status-cell"
          >
            <div
              aria-hidden="true"
              aria-label="available"
              class="available-marker"
              role="img"
            />
            Scheduled
          </div>
        </td>
      </tr>
      <tr
        class="file-table-row unselectable"
      >
        <td
          class="selected"
        >
          <button
            aria-checked="false"
            aria-label="Select"
            class="checkbox disabled"
            disabled=""
            role="checkbox"
          />
        </td>
        <td>
          7za.exe
        </td>
        <td>
          Toad
        </td>
        <td>
          \\Device\\HarddiskVolume1\\temp\\7za.exe
        </td>
        <td>
          <div
            class="status-cell"
          >
            <div
              aria-hidden="true"
              aria-label="available"
              class="available-marker"
              role="img"
            />
            Scheduled
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
